### ISSUE TITLE
Offer a way to edit xdg user dirs

### ISSUE THREAD

**Issue by jleclanche** [lxde/lxqt#780]
cc @luis-pereira 

https://wiki.freedesktop.org/www/Software/xdg-user-dirs/

We should offer a way to edit `$XDG_CONFIG_HOME/user-dirs.dirs`.

**Comment by jleclanche** [comment]
Suggest we add that in session settings for the time being.

**Comment by luis-pereira** [comment]
@jleclanche

```pcmanfm-qt --desktop``` sets the `Desktop` location. IMO doesn't make much sense having that in two places.

WIP screenshot
![screen-2015-08-17-08-29-41](https://cloud.githubusercontent.com/assets/62877/9308798/eddfc180-44ba-11e5-848b-2f2ff62baec8.png)

**Comment by jleclanche** [comment]
That looks good for the time being.

Regarding pcmanfm-qt I don't think it matters. Being able to edit a common setting in two places is OK, if both of them make sense.

**Comment by luis-pereira** [comment]
It matters from a performance perspective. I'm using libqtxdg and pcmanfm-qt uses internal stuff.  And it only allows to change the Desktop location.

What do you see missing / wrong in the screenshot ?

**Comment by jleclanche** [comment]
I think the UI can be simplified. I can't model what I have in mind right now but see if I can explain it.

Instead of a list like this, have a scrollable list view with the directories like our icons/cursor theme config. Each item has an icon representing it on the left (the icon that would be used for the directory), its name in bold, and the path in grey below. On the right, a single edit button to edit the current item, which sets its path.

Reduces the amount of items there.

As for pcmanfm, it should be use libqtxdg. For now we should leave the config option, we can always remove it later if we change our minds.

**Comment by pmattern** [comment]
Notes on layout / UI:

Cursor and icon theme are different as there's only one parameter and it's corresponding items to chose from. As for `xdg-user-dirs` there's one more layer in between: xdg-user-dirs as a whole - functionality like "Documents" - folder that's chosen.
Desktop environments are for end users the majority of which supposedly has never heard of `xdg-user-dirs` as a term. So I think an UI that's reflecting the hierarchy just depicted is pretty helpful to get an idea of the concept of `xdg-user-dirs` / the meaning of those directories.
On the other hand I guess an UI as proposed by jleclanche may be better for the experienced but rather confusing for the less experienced.

All in all I'd strongly vote for leaving the UI as shown in the screenshot which I for one find pretty good.

**Comment by Vladimir-csp** [comment]
+1 for what's on screenshot

**Comment by PCMan** [comment]
I think the UI mockup is good.
The problem is the confusion caused by the feature proposed.
For programs that are already running, they might cache the old values and keep using them.
Of course you can monitor the change of that xdg-user-dirs file and update the values in libqtxdg, but this will not work for programs not using libqtxdg and can create inconsistency between apps.
When should the changes be applied is a real problem and its behavior should be well defined.
Are the new values applied until next login, or they need to be applied immediately?
If we want to apply them immediately, how to keep the consistency among apps?
@luis-pereira It's OK to hide that option in the file manager while running in LXQt.

**Comment by pmattern** [comment]
Basically it would be desirable to have those settings applied at once to be consistent with all kinds of other LXQt GUIs handling settings like lxqt-config-appearance.
Don't know whether this is feasible regarding `xdg-user-dirs` which are normally only read when a session is launched to my knowledge.

**Comment by Vladimir-csp** [comment]
xdg_user_dirs are not even among environment variables, they are parsed from the config when application starts. So it is possible to change them inside session. And aware apps could get changes on the fly. As for the others, I think a warning message would suffice: "Some applications may not use the changes you've made until they restart. The best way to guarantiee that changes are applied by all applications is to restart your session."

**Comment by PCMan** [comment]
I still think that applying the changes all at once until next login is cleaner and more reliable. You don't change these settings everyday, do you? For this kind of one-time settings, having to logout and login again does not hurt usability much IMHO. On the other hand, trying to apply them on-the-fly creates many problems.

**Comment by Vladimir-csp** [comment]
xdg-user-dirs-update does this inside session.

**Comment by PCMan** [comment]
Quote from its doc: "The way it works is that xdg-user-dirs-update is run very early in the login phase."
It's not intended for updating things on-the-fly.
To my knowledge, all gtk+ programs are not aware of the changes.
See the source code of g_get_user_special_dir().
https://git.gnome.org/browse/glib/tree/glib/gutils.c#n1710
It caches the content of xdg-user-dirs settings unless you "manually" update the cache.
By default, it does not reload the config file automatically for obvious performance reasons.
So I'm against updating the settings on-the-fly since there is no reliable way to get it working as expected.

**Comment by luis-pereira** [comment]
@PCMan It's not a mockup. The code was implemented when I posted the screenshot. It needs minor enhancements. Let's clear some misunderstandings:
* It isn't an on the fly updating editor. Just and editor.
* It doesn't store anything inside libqtxdg. It reads/writes `user-dirs.dirs` . Check  [`QString XdgDirs::userDir(XdgDirs::UserDirectory dir)`](https://github.com/lxde/libqtxdg/blob/master/xdgdirs.cpp#L103) and [`bool XdgDirs::setUserDir(XdgDirs::UserDirectory dir, const QString& value, bool createDir)`](https://github.com/lxde/libqtxdg/blob/master/xdgdirs.cpp#L150)
* It needs an new login to make it work Ok. It warns the user about that.

Regarding the Ui: When designing it, I first consider something like what @jleclanche stated. But I found this one easier to understand.

I'm publishing the code. Keep in mind that's not finished yet. Check it [here](https://github.com/lxde/lxqt-session/pull/39)

**Comment by luis-pereira** [comment]
Update: Not a WIP anymore. Pls test.
@Vladimir-csp running `xdg-dirs-update` doens't solve the caching issue @PCMan refereed.

**Comment by jleclanche** [comment]
I still think the UI is too busy, but lgtm

**Comment by luis-pereira** [comment]
There is another consequence of merging https://github.com/lxde/lxqt-session/pull/39 and thus disable the desktop folder editing funcionality in ```pcmanf-qt --desktop```: 
* When the user changes the desktop folder, via lxqt-session, the desktop contents will not be updated until next login. 

Is this an acceptable behaviour ?

**Comment by jleclanche** [comment]
Hmm, no, that's not acceptable.

**Comment by Vladimir-csp** [comment]
Since $XDG_*_DIR are usually not exported to environment, pcmanfm (and other involved apps) could be triggered to re-read $XDG_CONFIG_HOME/user-dirs.dirs

**Comment by luis-pereira** [comment]
That's also my opinion, that's not acceptable. Although all other settings are only updated with a new login, the desktop folder at ```pcmanfm --desktop``` is a special case. 
A ```QFileSystemWatcher``` can be used to monitor ``` $XDG_CONFIG_HOME/user-dirs.dirs``` in this special case.

**Comment by Vladimir-csp** [comment]
On the other hand, updating this mid-session may be inconsistent with many other software. A note should be displayed anyway.

**Comment by luis-pereira** [comment]
@Vladimir-csp A note is always displayed.

**Comment by pmattern** [comment]
The actual feature is working without any issue but I think some polish could improve the UX:

Term "Locations" in left pane doesn't seem very telling to me.
Maybe just call this "User Directories" which would be more meaningful and point a little bit to the underlying spec / software as well?

The second issue is about the paths that are displayed.

The underlying [software / specification](http://freedesktop.org/wiki/Software/xdg-user-dirs/) enforces all corresponding folders to be set somehow and is automatically using `/home/<user>` if nothing particular is set. E. g. a situation "system doesn't know anything about folder *Music*" isn't possible, only "*Music* is set to `/home/<user>` instead of a dedicated folder right now". (This IMO not so obvious feature was recently discussed in #541, btw.)
Basically the UI in `lxqt-config-session` does reflect this. But clicking those icons `edit-clear-locationbar-rtl` at the right end of the forms providing the directories' paths will erase all content in the forms, path `/home/<user>` is only displayed when `lxqt-config-session` is launched the next time.

I think it would be better if clicking those icons showed path `/home/<user>` at once.
Also, it would be pretty good if the UI provided some information about the rationale to set `/home/<user>` and possibly the link stated above as well. I think information like this could be placed in the free space below the forms or made available in a dialog on its own that comes up by clicking a button `Help` or so.

**Comment by Vladimir-csp** [comment]
> ...which would be more meaningful and **point a little bit to the underlying spec / software** as well

Words of gold. 

But IMHO, there should be two buttons: disable and revert to default.

I may be wrong, but I would expect the 'clear' button to reset the path to default, because it is the only button (not counting browse...) and it has no tooltip, so I would assume the less destructive action is underneath.

There should be two buttons, with tooltips and clear visual response: 1. "Disable" leaves just the home dir in the path (the text in the field could be grayed out if it contains only home dir path to further indicate disabled state). 2. "Revert to default" does what it says.

Also the lack of separate 'Apply' in the settings dialog really bugs me.

**Comment by jleclanche** [comment]
+1 on User Directories.

**Comment by pmattern** [comment]
Given the wicked meaning "disabling" gets assigned by the specification I'd tend to avoid naming an eventual button setting the paths to `/home/<user>` and/or a corresponding tooltip like this. Rather, I'd label these e. g. "Point to /home/\<user\>".
In conjunction with some eventual notes about the specification or the link stated above this could help users understand the concept better and avoid confusion due to that special meaning of "disable".

> Also the lack of separate 'Apply' in the settings dialog really bugs me.

This is a bit surprising.
The majority of settings in `lxqt-config-session` gets valid upon the next login only anyway and in the discussion above it was agreed that it's best to handle those xdg-user-dir directories the same way.
Also, the vast majority of settings like all from `lxqt-config-appearance` are applied right upon selection so an `Apply` button doesn't exist. This behaviour may be questionable but it's ubiquitous in LXQt right now so the lack of an `Apply` button in the section dealing with xdg-user-dirs in `lxqt-config-session` is just consistent.

**Comment by Vladimir-csp** [comment]
"Point to /home/$USER
(effectively disable)"

Doing things without apply button just feels like walking in a mindefiled. The ability to just close dialog (or hit Cancel) and forget about the whole deal comes really handy sometimes. Without it any screwup needs to be fixed by hand. That is not limited to session config of course.

**Comment by luis-pereira** [comment]
@Vladimir-csp The Apply button should be deployed on all LXQt, not just here. That's an issue to be addressed at `LXQt::ConfigDialog`
Putting a link to a specification in the dialog will not be of much help to most users IMO. A detailed explanation belongs to an Help system..... that we don't have... yet.

@pmattern I'm sympathetic to your arguments over the clear button. Isn't clicking in the clear button and not having the text cleared misleading in some way ?

IMO what we really need is an 'Defaults' button in `LXQt::ConfigDialog`

What kind of tooltip do you suggest ?

**Comment by pmattern** [comment]
First of all I've stumbled upon another problem the outcome of which could have an impact on what we're discussing here, see #896.
If the decision should be made that section "Locations" should be usable when binaries `xdg-user-dir[s-update]` are absent as well the whole topic "disable vs. `/home/<user>/`" needed review.

That aside it's of course confusing and should be avoided that a button that's well-known to clean a form is doing something different all of a sudden.
But I don't know whether buttons similar to "clear" as provided right now (we do talk about this one,
![lxqt-config-session_xdg-user-dirs_button-clear](https://cloud.githubusercontent.com/assets/9860801/11168591/e7ddf532-8b95-11e5-82a3-4c457f597f33.png)
right?) but providing different features like setting a particular directory can get implemented. Also, I think having both a button to set path `/home/<user>/` and another one to set the default path in terms of xdg-user-dirs wouldn't be bad.
But wouldn't two buttons like this in a single form, if technically feasible at all, be bad to handle? Should the UI eventually be revamped completely? How?
Really don't know what's the best thing to do here.

Tooltips weren't proposed by me. Nonetheless I think they could very well be useful but the details depend on the outcome of all the rest.

Topic `Apply` button really shouldn't be discussed here, IMO.
Personally I agree that the current policy to apply all kinds of settings right when they're chosen isn't convenient to handle and that it would be much better to separate making settings from applying them via a dedicated button, much like it's handled by KDE's System Settings.
But this is a rather major design decision that should be treated as such and hence not here, maybe in #893.

**Comment by luis-pereira** [comment]
@pmattern 
What about something like that: 
![screen-2015-11-24-18-32-00](https://cloud.githubusercontent.com/assets/62877/11376627/a763b690-92d9-11e5-9346-ff3505284fd0.jpg)

A mockup of https://github.com/lxde/lxqt/issues/893. It solves the pending issues here IMO.

**Comment by pmattern** [comment]
Basically I'd be fine with that.

Cannot guess the actions performed by the buttons that got added from the screenshot.
I'm assuming `Reset` → "restore former values" and `Restore Defaults` → set the default directories like `/home/{Desktop,Downloads,Documents,...,Pictures,...}` or their translated counterparts. And what about the icons at the forms' right ends? Would they still clear the forms leaving the problems depicted above ("transformation into `/home/<user>` upon exit") as is? Or would / could these icons set `/home/<user>` at once?
Regarding `Help` I don't know. Would it be some background information about xdg-user-dirs? Personally, I still think we should provide some hints on that wicked "disable = set to `/home/<user>`" thing somehow.

Is button `Close` missing in the mockup only or do you intend to drop it?

On a side note, what's the matter with "Music" in the screenshot? Something similar as depicted in #895?

**Comment by Vladimir-csp** [comment]
IMHO 'clear' widget inside text field does not fit here. Because it does not clear the text field. And clearing the text field has no sense in this situation, so it is better not to nudge users into doing so.

There should be three buttons next to each field
(with tooltips):
- Reset
- Set to $HOME (effectively disable)
- Browse

(or Reset could be common as in the mockup, which leaves two buttons next to field.)

**Comment by pmattern** [comment]
@luis-pereira ad https://github.com/lxde/lxqt-session/pull/57#issuecomment-197569455:
→ lxde/lxqt-session@8f653bd Obviously alright. Didn't make it to `./lxqt-config-session/translations` yet but this seems to be due to the fact that the directory hasn't been updated since.
→ lxde/lxqt-session@01d62e5 IMHO the English has got some rough edges but not 100% sure how to put it better, maybe we'll get feedback of a native speaker some day. But if I'm not mistaken the tooltips aren't subject to translation yet. If so this should be complemented.
→ lxde/lxqt-session@1e7f82c Tbh. I couldn't see any effect of this commit so far.

That aside I still do not think that "clear" button at the forms' right end makes sense for the reasons discussed extensively above.

**Comment by luis-pereira** [comment]
@pmattern Without https://github.com/lxde/lxqt-session/commit/1e7f82c if the user closed the dialog with an empty entry, the working directory would be set as the chosen dir

**Comment by pmattern** [comment]
@luis-pereira 
I see, if `lxqt-config-session` gets launched from a terminal emulator, right? Thanks for clarifying.
